src/Features/LanguageServer/Protocol/CustomProtocol/FindUsagesLSPContext.cs

```diff
@@ -250,28 +255,79 @@ public override async ValueTask OnReferenceFoundAsync(SourceReferenceItem refere
                 }
             }
 
-            static async Task<object?> ComputeTextAsync(
+            static async Task<ClassifiedTextElement?> ComputeTextAsync(
                 int id, int? definitionId,
                 DocumentSpan documentSpan,
                 ClassifiedTextElement? definitionText,
+                bool isWrittenTo,
                 CancellationToken cancellationToken)
             {
-                if (id == definitionId)
-                {
-                    return definitionText;
-                }
-                else if (documentSpan != default)
+                // General case
+                if (documentSpan != default)
                 {
                     var classifiedSpansAndHighlightSpan = await ClassifiedSpansAndHighlightSpanFactory.ClassifyAsync(
                         documentSpan, cancellationToken).ConfigureAwait(false);
                     var classifiedSpans = classifiedSpansAndHighlightSpan.ClassifiedSpans;
                     var docText = await documentSpan.Document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    var classifiedTextRuns = GetClassifiedTextRuns(id, definitionId, documentSpan, isWrittenTo, classifiedSpans, docText);
 
-                    return new ClassifiedTextElement(
-                        classifiedSpans.Select(cspan => new ClassifiedTextRun(cspan.ClassificationType, docText.ToString(cspan.TextSpan))));
+                    return new ClassifiedTextElement(classifiedTextRuns.ToArray());
+                }
+                // Certain definitions may not have a DocumentSpan, such as namespace and metadata definitions
+                else if (id == definitionId)
+                {
+                    return definitionText;
                 }
 
                 return null;
+
+                // Nested local functions
+                static ClassifiedTextRun[] GetClassifiedTextRuns(
+                    int id, int? definitionId,
+                    DocumentSpan documentSpan,
+                    bool isWrittenTo,
+                    ImmutableArray<ClassifiedSpan> classifiedSpans,
+                    SourceText docText)
+                {
+                    using var _ = ArrayBuilder<ClassifiedTextRun>.GetInstance(out var classifiedTextRuns);
+                    foreach (var span in classifiedSpans)
```

Comment:
one thing i'm curious about.  if we have something static, don't we get two classifications for the same span?  what happens then in LSP?  can you just validate that there are no issues with static defs/refs (and add test)?

Reply:
Interestingly enough, `ClassifiedSpansAndHighlightSpanFactory.ClassifyAsync` call eventually [consolidates duplicate spans](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Classification/ClassifierHelper.cs,52), so this doesn't seem to be an issue at present. I will add a test to make sure any changes don't break LSP in the future though, thank you!